client/src/index.ts

```diff
@@ -0,0 +1 @@
+export * from './types'
```

Comment:
I would prefer to not export types in from `/client/src/index.ts` and `/index.ts` to not mix types package and PeerTube main code. IMHO types is just an additional package, so we should not have them as the main TS entries.

Could we imagine creating a `/types/` or `/packages/types` directory containing these index files?
Then we can put compiled files in `/types/dist` or `/packages/types/dist` because the current `/types` directory could be confusing (is it a build directory or a source directory?).

Reply:
From what I understand, in the client codebase, what you're putting in `client/src/types/` aren't external types but rather __PeerTube__ client own *interfaces*.
As __types__ and  __typings__  are usually used as synonymous, I would rather not put anything part of the __PeerTube__ codebase inside this directory. Ideally I would only put *missing* external typings.

For example on the `server` I find it rather confusing to find:
- `server/models` :heavy_check_mark:
- `server/typings` :heavy_check_mark:
- `server/types` which is containing __domain__ related `models` (`server/types/models` for example)
I'd rather merge `server/models` and`server/typings` directories.

...after some thinking I think I understand why this directory is named `types`, it contains everything declared with the `type` keyword, is that right? Might sounds silly but wasn't obvious to me... :sweat_smile:  
Is there another rational behind this that I'm not getting?

I still think most of these `types` could live close to their related model. 
I'm having a look at the `Actor` model right now... Maybe `MActor`, `MActorAccountChannelId`, etc. `types` could live in the same file? Or same dir? Just thinking out loud here. I wonder how other projects are doing.
By the way, what's the `M` of `MActor` is standing for?

> *Just a few interesting resources I browsed while wondering if some of the types could have been interfaces.*
> https://stackoverflow.com/questions/37233735/interfaces-vs-types-in-typescript/52682220#52682220
> https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces

Let's get back to our issue, 

> IMHO types is just an additional package, so we should not have them as the main TS entries.

IMHO this `@peertube/peertube-types` package shouldn't exist at all but `@peertube/peertube` should instead (or `@peertube/*` with the different packages).
It only exist because I didn't want to mess the current codebase too much and at the moment `types` are needed for third parties developments.
I think that a `@peertube/peertube` should be set as a mid to long term target, and that doing as such, it would help drive future code organization choices.

I wonder... is this `types` directory exposing only publicly visible types?
https://www.typescriptlang.org/tsconfig#stripInternal
Could the `@peertube/peertube-types` only exposing the different `types` directories and expose everything needed?
If so, it fails as doing so, but I might getting ahead of myself here.

> Could we imagine creating a /types/ or /packages/types directory containing these index files?
> Then we can put compiled files in /types/dist or /packages/types/dist because the current /types directory could be confusing (is it a build directory or a source directory?).

Sure, I'll play with this idea a bit and see where I'm getting! :+1:

Reply:
> From what I understand, in the client codebase, what you're putting in client/src/types/ aren't external types but rather PeerTube client own interfaces.

Yes, this directory is dedicated to peertube client own interfaces. The `types` naming is not very good (same as `server/types`, that contains PeerTube server own interfaces).

> Is there another rational behind this that I'm not getting?

No, it's just a bad named directory :) I'm open to suggestions (but I'm not sure to understand why you'd "merge server/models andserver/typings directories")

>  I wonder how other projects are doing.

I never found other typescript projects that tries to type DB associations.

> By the way, what's the M of MActor is standing for?

`Model` I guess, it's just to clearly differentiate `ActorModel` (with all optional associations) than `MActor` (typed associations).

> Just a few interesting resources I browsed while wondering if some of the types could have been interfaces.

Since we use `Omit`, `Pick` etc I preferred to keep consistency with `type`. If you find a clearer way, with interfaces to describe model associations don't hesitate to explain it :)

> I think that a @peertube/peertube should be set as a mid to long term target, and that doing so, would help drive future code organization choices.

I don't understand understand the point of publishing a `@peertube/peertube` package. We don't plan to release PeerTube on NPM, so why would we publish it? Instead, I always thought we would publish packages in `@peertube/peertube-*` that could be used by other libraries (`types` of course, but we could also consider to release a `@peertube/peertube-server-helpers` package for example so other projects can use them). But I may miss your point?

> I wonder... is this types directory exposing only publicly visible types?

I'm not sure to understand what `types` directory you're talking about.

Reply:
> No, it's just a bad named directory :) I'm open to suggestions (but I'm not sure to understand why you'd "merge server/models and server/typings directories")

No worries :D
I meant to merge `server/models` and `server/types/*` or even move some of `server/types/models/*` inside their related model files. It seems most of them aren't used -- mostly -- anywhere else. Just planting seeds here :seedling:

> We don't plan to release PeerTube on NPM, so why would we publish it

Well if it's not planned, I understand.
But instead of a `@peertube/peertube-server-helpers` I was rather thinking about a standalone `@peertube/server` for example. 
> As I was saying in previous comments it could have provided the types, the code and a standalone __PeerTube__ server at the same time. There wouldn't be the need to create a `*-types` package.

I guess what I'm saying is, reasoning in *packages*, helps to set clearer boundaries between each __apps__.

> I wonder... is this types directory exposing only publicly visible types?
> >I'm not sure to understand what types directory you're talking about.

I think I misunderstand one of your point but... I was saying that if **all** of the `types` (server, client, etc.) directories were exposing *every* interfaces required by third parties, it might make more sense to me to exposes them and only them in a `*types` package.

Reply:
> I meant to merge server/models and server/types/* or even move some of server/types/models/* inside their related model files. It seems most of them aren't used -- mostly -- anywhere else. Just planting seeds here seedling

I think it's a good idea :+1:

>  I was rather thinking about a standalone @peertube/server for example.

And you'd put every server module in this package?